tensorflow/core/kernels/sparse_tensor_dense_matmul_op.cc

```diff
@@ -132,22 +147,39 @@ class SparseTensorDenseMatMulOp : public OpKernel {
       return;
     }
 
-#define MAYBE_ADJOINT(ADJ_A, ADJ_B)                                        \
-  if (adjoint_a_ == ADJ_A && adjoint_b_ == ADJ_B) {                        \
-    Status functor_status = functor::SparseTensorDenseMatMulFunctor<       \
-        Device, T, Tindices, ADJ_A,                                        \
-        ADJ_B>::Compute(ctx->eigen_device<Device>(), out->matrix<T>(),     \
-                        a_indices->matrix<Tindices>(), a_values->vec<T>(), \
-                        b->matrix<T>());                                   \
-    OP_REQUIRES_OK(ctx, functor_status);                                   \
+    switch (ndim) {
+#define MAYBE_ADJOINT(ADJ_A, ADJ_B, N)                               \
```

Comment:
General comment: Adding specializations and templating on more and more parameters bloats the binary code size quite significantly. We should avoid it and pass ndims (rank) as a runtime parameter wherever possible.

Reply:
@zycdragonball I would recommend following the recommendation of ebrevdo@ here. You can see other example of this pattern, https://github.com/tensorflow/tensorflow/blob/64f0ebd33a7c868da3c8f1ea15adf358c578f227/tensorflow/core/kernels/matrix_inverse_op.cc#L137